scanner/decisions/0001-scannerv4-apis.md

```diff
@@ -0,0 +1,93 @@
+2. `scanner.v4.Indexer/GetIndex`: Retrieve or check index existence. Returns [`IndexReport`](https://github.com/quay/claircore/blob/v1.4.18/indexreport.go#L19).
+3. `scanner.v4.Matcher/GetVulnerabilities`: Get vulnerabilities for a given resource's manifest. Returns [`VulnerabilityReport`](https://github.com/quay/claircore/blob/v1.4.18/vulnerabilityreport.go#L7).
+4. `scanner.v4.Matcher/GetVulnerabilityMetadata`: Get information on vulnerability metadata, e.g., last update timestamp.
+
+Example of manifest's `hash_id` usage to create index reports for container images:
+
+```
+message ContainerImageLocator {
+    string url      = 1;
+    string username = 2;
+    string password = 3;
+}
+
+message CreateIndexRequest {
+    string hash_id = 1;
+    oneof resource_locator {
+        ContainerImageLocator container_image;
+    }
+}
+```
+
+Example of index request and vulnerability request:
+
+```
+message GetIndexRequest {
+    string hash_id   = 1;
+    // If true, does not return the Index Report, only status.
+    bool check_only  = 2;
+}
+
+message GetVulnerabilitiesRequest {
+    string hash_id = 1;
+}
+```
+
+Examples of common error codes and scenarios:
+
+| Code | Scenario |
+| --- | ---
+| `UNAVAILABLE` | Scanner temporary or retriable issue caused by dependencies (e.g., registries, databases).
+| `INVALID_ARGUMENT` | Invalid input or malformed input (e.g., bad credentials, malformed `hash_id`).
+| `INTERNAL` | Unrecoverable errors, unexpected conditions, or invariant (e.g., bugs, broken schema, etc).
+| `FAILED_PRECONDITION	` | Index report does not exist for this given `hash_id` (e.g. when calling `GetVulnerabilities`).
+
+All APIs are idempotent, hence retriable on timeouts and temporary errors.
+
+## Consequences
+
+Leveraging gRPC-only helps maintain service contracts and enhances performance. It also aligns with other Stackrox services.
+
+The new image integration for ScannerV4 ensures minimal changes are necessary for Central to switch between ScannerV2 and ScannerV4 using a feature flag.
+
+The Indexer and Matcher modes allow additional deployment options, enabling ScannerV4 to scale and serve different scenarios. The multiple modes and gRPC services pave the way for advanced deployments, such as multi-tenant scanning and node scanning.
+
+Using gRPC exclusively increases complexity for load balancing ScannerV4 for horizontal scalability. Additional steps will be necessary to achieve this, including service-mesh tooling or client-side load-balancing, which are not available out of the box and require additional work.
+
+The client-managed `hash_id` approach for identifying manifests may lead to duplicated index reports for locators pointing to the same resource. While this simplifies the API to support future resource types (orchestrator, nodes), it may lead to client misuse.
```

Comment:
You'll want to make sure to resolve the image location as soon as possible and then use the manifest and layer digests to avoid duplicate work.

Reply:
Agreed. We are also looking into ensuring we can avoid multiple round-trips to the registries when creating the manifest. CC @RTann

Reply:
The scanner needs to support not only container image manifests but, in the future, node manifests, etc. Hence, the APIs assume the ID is managed outside of Scanner. That allows the client to predictability generate IDs, and call APIs without having to retrieve it, but opens the door for misuse -- two calls with the same image locator and two different IDs. Or, the same ID for two different image locators.